pkg/util/future/future.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package future
+
+import "sync/atomic"
+
+// Future represents a value which will become ready in the future.
+// A zero initialized Future is ready to be used, and it represents a promise
+// that the caller will produce value in the future.
+// The caller may arrange for a callback to be invoked when the future
+// becomes ready.
+// Use AwaitableFuture if the caller needs to wait for the future.
+type Future[T any] struct {
+	value atomic.Pointer[futureValue[T]]
+}
+
+// futureValue holds value, along with any callbacks that
+// need to run when value becomes available.
+type futureValue[T any] struct {
+	v         *T
+	whenReady []func(v T)
+}
+
+// ErrorFuture is a handy alias for a future that returns an error.
+type ErrorFuture = Future[error]
+
+// AwaitableFuture is a future that can be waited on.
+type AwaitableFuture[T any] struct {
+	*Future[T]
+	done <-chan struct{}
+}
+
+// MakeCompletedFuture constructs future which returns specified value.
+func MakeCompletedFuture[T any](v T) *Future[T] {
+	f := MakePromise[T]()
+	f.Set(v)
+	return f
+}
+
+// MakePromise constructs future where the caller promises to set the value or
+// an error.
+func MakePromise[T any]() *Future[T] {
+	return &Future[T]{}
+}
+
+// MakeCompletedErrorFuture constructs completed ErrorFuture.
+func MakeCompletedErrorFuture(err error) *ErrorFuture {
+	f := ErrorFuture{}
+	f.Set(err)
+	return &f
+}
+
+// MakeAwaitableFuture wraps underlying future so that the callers can wait
+// for future to become available.
+func MakeAwaitableFuture[T any](f *Future[T]) AwaitableFuture[T] {
```

Comment:
Yes; I actually had that done early on.  I happen to really like this pattern.
One of the reasons is that to immediately make awaitable future, you have to construct a channel.  That's not super cheap.  Alternatively, you do something like context does (which Is what I had early) -- and that is, construct channel lazily, but use mutex to synchronize.  That's also not cheap.

I think awaitable future strikes a very good balance.  For one, some users of Future do not want to wait for it -- they just want to arrange for WhenReady to execute when future is ready.

Only the clients that want to wait for future would construct AwaitableFuture, and if the underlying future happens to be completed already (which is fairly common, particularly when using ErrorFuture), the construction is very cheap (since closed channel is used).